js/src/forum/addPushNotifications.js

```diff
@@ -172,4 +207,51 @@ export default () => {
       );
     }
   });
+
+  extend(SettingsPage.prototype, 'notificationsItems', function (items) {
+    if (!usingAppleWebview()) return;
+
+    if (firebasePushNotificationState !== 'authorized') {
+      items.add(
+        'firebase-push-optin-default',
+        Alert.component(
+          {
+            itemClassName: 'pwa-setting-alert',
+            dismissible: false,
+            controls: [
+              Button.component(
+                {
+                  className: 'Button Button--link',
+                  onclick: () => {
+                    window.webkit.messageHandlers['push-permission-request'].postMessage('push-permission-request');
+                  },
+                },
+                app.translator.trans('askvortsov-pwa.forum.settings.pwa_notifications.access_default_button') + firebasePushNotificationState
+              ),
+            ],
+          },
+          [icon('fas fa-exclamation-circle'), app.translator.trans('askvortsov-pwa.forum.settings.pwa_notifications.access_default')]
+        ),
+        10
+      );
+    }
+  });
+
+  extend(SettingsPage.prototype, 'oncreate', function () {
+    if (usingAppleWebview()) {
+      window.webkit.messageHandlers['push-permission-state'].postMessage('push-permission-state');
+
+      window.addEventListener('push-permission-request', requestFirebasePushNotificationPermission);
```

Comment:
Where do these events come from? I can't find documentation for them on mdn

Edit: I assume they're from the webkit messageHandlers. Could you put the string constants in a variable so we don't need to worry about keeping them in sync, and add a comment explaining that webkit messageHandlers will generate custom Dom events?

Reply:
When using webview, it seems that's how communication happens between Swift and the Webview. This allows us to request state from Swift and have swift dispatch things back to the webview.

https://developer.apple.com/documentation/webkit/wkscriptmessagehandler

The events themselves are the default events provided by PWABuilder, so it's specific to the default template PWABuilder generates. (I can't find exactly their repo with the swift code in it though)

I'll move them up so it's easier to update them and reference them. 👍